### PR TITLE
Aseprite atlases in JSON format for FlxAtlasFrames

### PR DESCRIPTION
**Pull request by MAJigsaw77**
it think the title says all

### CODE REVIEW

**Comment by Geokureli** [comment]
I believe aseprite outputs a json matching texture packers format, so this is already supported

**Comment by Geokureli** [comment]
I'm under the impression that ASEprites json output is the same as TexturePacker's, so this might already be implemented by flixel

**Comment by MAJigsaw77** [comment]
They are different, aren't the same, this should be a good addition but let me fix the compile

**Comment by MAJigsaw77** [comment]
Also just a little thing, this is based on TexturePacker's one lol

**Comment by MAJigsaw77** [comment]
@Geokureli sorry if you are busy but can you give approval for the workflow?

**Comment by MAJigsaw77** [comment]
Now should be fixed

File: flixel/graphics/frames/FlxAtlasFrames.hx [hunk 1]
```diff
@@ -421,3 +509,93 @@ typedef TexturePackerObject =
 {
 	frames:Dynamic
 }
+
+typedef AsepriteData =
+{
+	var frames:Dynamic,
```
Comment (Geokureli) [inline flixel/graphics/frames/FlxAtlasFrames.hx:515]: this is the issue, this should have a ;

**Comment by Geokureli** [comment]
are you running this locally before committing?

File: flixel/graphics/frames/FlxAtlasFrames.hx (hunk 1 above)
Comment (Geokureli) [inline flixel/graphics/frames/FlxAtlasFrames.hx:515]: i think typdefs either need to be one of the following
```hx
{ foo:String, bar:String}
```
or
```hx
{ var foo:String; var bar:String; }
```
notice the ;

try to use your branch of flixel in a test project to make sure it compiles

**Comment by MAJigsaw77** [comment]
> 

I have issues with the compile literally, my pc is shit

File: flixel/graphics/frames/FlxAtlasFrames.hx (hunk 1 above)
Comment (MAJigsaw77) [reply flixel/graphics/frames/FlxAtlasFrames.hx:515]: Ok

**Comment by Geokureli** [comment]
compilations work, gonna test some things out locally

**Comment by MAJigsaw77** [comment]
Ok

**Comment by Geokureli** [comment]
Maybe I'm misunderstanding the purpose of this but I'm still not sure if this is even necessary, I was able to load an aseprite json into flixel with the following code:
```hx
var spr = new FlxSprite();
spr.frames = FlxAtlasFrames.fromTexturePackerJson("assets/alien-sheet.png", "assets/alien.json");
spr.animation.addByPrefix("white", "white", 4);
spr.animation.addByPrefix("red", "red", 4);
spr.animation.play("white");
add(spr);
```
using the following export settings:
Screen Shot 2022-07-22 at 2 35 59 PM
Screen Shot 2022-07-22 at 2 36 08 PM

**Comment by MAJigsaw77** [comment]
You need to be sure you put tag's on all the animations and to select the sprite sheet type to packed and at the Output put `{title} {tag}000{tagframe}` also you put the wrong code for sprite sheet

**Comment by MAJigsaw77** [comment]
This is how should be
```
var spr = new FlxSprite();
spr.frames = FlxAtlasFrames.fromAseprite("assets/alien-sheet.png", "assets/alien.json");
spr.animation.addByPrefix("white", "white", 4);
spr.animation.addByPrefix("red", "red", 4);
spr.animation.play("white");
add(spr);
```

**Comment by MAJigsaw77** [comment]
@Geokureli

**Comment by MAJigsaw77** [comment]
> Maybe I'm misunderstanding the purpose of this but I'm still not sure if this is even necessary, I was able to load an aseprite json into flixel with the following code:
> 
> ```hx
> var spr = new FlxSprite();
> spr.frames = FlxAtlasFrames.fromTexturePackerJson("assets/alien-sheet.png", "assets/alien.json");
> spr.animation.addByPrefix("white", "white", 4);
> spr.animation.addByPrefix("red", "red", 4);
> spr.animation.play("white");
> add(spr);
> ```
> 
> using the following export settings: Screen Shot 2022-07-22 at 2 35 59 PM Screen Shot 2022-07-22 at 2 36 08 PM
> 
> Can you be specific about what this new method allows that wasn't possible before?

Btw TexturePacker atlas is different then the aseprite one

**Comment by Geokureli** [comment]
yes you said that, but I don't see how they are different, can you explain the difference?

**Comment by MAJigsaw77** [comment]
look

this one is texurePacher
```json
{
	"frames":{
		"discord.png":{
			"frame":{
				"x":0,
				"y":0,
				"w":100,
				"h":100
			},
			"rotated":false,
			"trimmed":false,
			"spriteSourceSize":{
				"x":0,
				"y":0,
				"w":100,
				"h":100
			},
			"sourceSize":{
				"w":100,
				"h":100
			}
		},
		"gamebanana.png":{
			"frame":{
				"x":100,
				"y":0,
				"w":100,
				"h":100
			},
			"rotated":false,
			"trimmed":false,
			"spriteSourceSize":{
				"x":0,
				"y":0,
				"w":100,
				"h":100
			},
			"sourceSize":{
				"w":100,
				"h":100
			}
		},
		"roblox.png":{
			"frame":{
				"x":200,
				"y":0,
				"w":100,
				"h":100
			},
			"rotated":false,
			"trimmed":false,
			"spriteSourceSize":{
				"x":0,
				"y":0,
				"w":100,
				"h":100
			},
			"sourceSize":{
				"w":100,
				"h":100
			}
		},
		"soundcloud.png":{
			"frame":{
				"x":300,
				"y":0,
				"w":100,
				"h":100
			},
			"rotated":false,
			"trimmed":false,
			"spriteSourceSize":{
				"x":0,
				"y":0,
				"w":100,
				"h":100
			},
			"sourceSize":{
				"w":100,
				"h":100
			}
		},
		"twitch.png":{
			"frame":{
				"x":400,
				"y":0,
				"w":100,
				"h":100
			},
			"rotated":false,
			"trimmed":false,
			"spriteSourceSize":{
				"x":0,
				"y":0,
				"w":100,
				"h":100
			},
			"sourceSize":{
				"w":100,
				"h":100
			}
		},
		"twitter.png":{
			"frame":{
				"x":500,
				"y":0,
				"w":100,
				"h":100
			},
			"rotated":false,
			"trimmed":false,
			"spriteSourceSize":{
				"x":0,
				"y":0,
				"w":100,
				"h":100
			},
			"sourceSize":{
				"w":100,
				"h":100
			}
		},
		"youtube.png":{
			"frame":{
				"x":600,
				"y":0,
				"w":100,
				"h":100
			},
			"rotated":false,
			"trimmed":false,
			"spriteSourceSize":{
				"x":0,
				"y":0,
				"w":100,
				"h":100
			},
			"sourceSize":{
				"w":100,
				"h":100
			}
		}
	},
	"meta":{
		"app":"https://www.codeandweb.com/texturepacker",
		"version":"1.0",
		"image":"icons.png",
		"format":"RGBA8888",
		"size":{
			"w":700,
			"h":100
		},
		"scale":"1",
		"smartupdate":"$TexturePacker:SmartUpdate:e3a7a53be6bbd88ae66e3d1f25e5c96a:5001547021c8d17d6e7757a752048662:b5c5bd7252d2048fecbb1bda1760f9ef$"
	}
}
```

this one is aseprite
```json
{
   "frames":{
      "bfPixel DOWN MISS 0000":{
         "frame":{
            "x":0,
            "y":0,
            "w":75,
            "h":75
         },
         "rotated":false,
         "trimmed":false,
         "spriteSourceSize":{
            "x":0,
            "y":0,
            "w":75,
            "h":75
         },
         "sourceSize":{
            "w":75,
            "h":75
         },
         "duration":100
      },
      "bfPixel DOWN MISS 0001":{
         "frame":{
            "x":75,
            "y":0,
            "w":75,
            "h":75
         },
         "rotated":false,
         "trimmed":false,
         "spriteSourceSize":{
            "x":0,
            "y":0,
            "w":75,
            "h":75
         },
         "sourceSize":{
            "w":75,
            "h":75
         },
         "duration":100
      },
      "bfPixel DOWN 0000":{
         "frame":{
            "x":150,
            "y":0,
            "w":75,
            "h":75
         },
         "rotated":false,
         "trimmed":false,
         "spriteSourceSize":{
            "x":0,
            "y":0,
            "w":75,
            "h":75
         },
         "sourceSize":{
            "w":75,
            "h":75
         },
         "duration":100
      },
      "bfPixel DOWN 0001":{
         "frame":{
            "x":225,
            "y":0,
            "w":75,
            "h":75
         },
         "rotated":false,
         "trimmed":false,
         "spriteSourceSize":{
            "x":0,
            "y":0,
            "w":75,
            "h":75
         },
         "sourceSize":{
            "w":75,
            "h":75
         },
         "duration":100
      },
      "bfPixel IDLE 0000":{
         "frame":{
            "x":300,
            "y":0,
            "w":75,
            "h":75
         },
         "rotated":false,
         "trimmed":false,
         "spriteSourceSize":{
            "x":0,
            "y":0,
            "w":75,
            "h":75
         },
         "sourceSize":{
            "w":75,
            "h":75
         },
         "duration":100
      },
      "bfPixel IDLE 0001":{
         "frame":{
            "x":375,
            "y":0,
            "w":75,
            "h":75
         },
         "rotated":false,
         "trimmed":false,
         "spriteSourceSize":{
            "x":0,
            "y":0,
            "w":75,
            "h":75
         },
         "sourceSize":{
            "w":75,
            "h":75
         },
         "duration":100
      },
      "bfPixel IDLE 0002":{
         "frame":{
            "x":450,
            "y":0,
            "w":75,
            "h":75
         },
         "rotated":false,
         "trimmed":false,
         "spriteSourceSize":{
            "x":0,
            "y":0,
            "w":75,
            "h":75
         },
         "sourceSize":{
            "w":75,
            "h":75
         },
         "duration":100
      },
      "bfPixel IDLE 0003":{
         "frame":{
            "x":525,
            "y":0,
            "w":75,
            "h":75
         },
         "rotated":false,
         "trimmed":false,
         "spriteSourceSize":{
            "x":0,
            "y":0,
            "w":75,
            "h":75
         },
         "sourceSize":{
            "w":75,
            "h":75
         },
         "duration":100
      },
      "bfPixel IDLE 0004":{
         "frame":{
            "x":600,
            "y":0,
            "w":75,
            "h":75
         },
         "rotated":false,
         "trimmed":false,
         "spriteSourceSize":{
            "x":0,
            "y":0,
            "w":75,
            "h":75
         },
         "sourceSize":{
            "w":75,
            "h":75
         },
         "duration":100
      },
      "bfPixel LEFT MISS 0000":{
         "frame":{
            "x":675,
            "y":0,
            "w":75,
            "h":75
         },
         "rotated":false,
         "trimmed":false,
         "spriteSourceSize":{
            "x":0,
            "y":0,
            "w":75,
            "h":75
         },
         "sourceSize":{
            "w":75,
            "h":75
         },
         "duration":100
      },
      "bfPixel LEFT MISS 0001":{
         "frame":{
            "x":750,
            "y":0,
            "w":75,
            "h":75
         },
         "rotated":false,
         "trimmed":false,
         "spriteSourceSize":{
            "x":0,
            "y":0,
            "w":75,
            "h":75
         },
         "sourceSize":{
            "w":75,
            "h":75
         },
         "duration":100
      },
      "bfPixel LEFT 0000":{
         "frame":{
            "x":825,
            "y":0,
            "w":75,
            "h":75
         },
         "rotated":false,
         "trimmed":false,
         "spriteSourceSize":{
            "x":0,
            "y":0,
            "w":75,
            "h":75
         },
         "sourceSize":{
            "w":75,
            "h":75
         },
         "duration":100
      },
      "bfPixel LEFT 0001":{
         "frame":{
            "x":900,
            "y":0,
            "w":75,
            "h":75
         },
         "rotated":false,
         "trimmed":false,
         "spriteSourceSize":{
            "x":0,
            "y":0,
            "w":75,
            "h":75
         },
         "sourceSize":{
            "w":75,
            "h":75
         },
         "duration":100
      },
      "bfPixel RIGHT MISS 0000":{
         "frame":{
            "x":975,
            "y":0,
            "w":75,
            "h":75
         },
         "rotated":false,
         "trimmed":false,
         "spriteSourceSize":{
            "x":0,
            "y":0,
            "w":75,
            "h":75
         },
         "sourceSize":{
            "w":75,
            "h":75
         },
         "duration":100
      },
      "bfPixel RIGHT MISS 0001":{
         "frame":{
            "x":1050,
            "y":0,
            "w":75,
            "h":75
         },
         "rotated":false,
         "trimmed":false,
         "spriteSourceSize":{
            "x":0,
            "y":0,
            "w":75,
            "h":75
         },
         "sourceSize":{
            "w":75,
            "h":75
         },
         "duration":100
      },
      "bfPixel RIGHT 0000":{
         "frame":{
            "x":1125,
            "y":0,
            "w":75,
            "h":75
         },
         "rotated":false,
         "trimmed":false,
         "spriteSourceSize":{
            "x":0,
            "y":0,
            "w":75,
            "h":75
         },
         "sourceSize":{
            "w":75,
            "h":75
         },
         "duration":100
      },
      "bfPixel RIGHT 0001":{
         "frame":{
            "x":1200,
            "y":0,
            "w":75,
            "h":75
         },
         "rotated":false,
         "trimmed":false,
         "spriteSourceSize":{
            "x":0,
            "y":0,
            "w":75,
            "h":75
         },
         "sourceSize":{
            "w":75,
            "h":75
         },
         "duration":100
      },
      "bfPixel UP MISS 0000":{
         "frame":{
            "x":1275,
            "y":0,
            "w":75,
            "h":75
         },
         "rotated":false,
         "trimmed":false,
         "spriteSourceSize":{
            "x":0,
            "y":0,
            "w":75,
            "h":75
         },
         "sourceSize":{
            "w":75,
            "h":75
         },
         "duration":100
      },
      "bfPixel UP MISS 0001":{
         "frame":{
            "x":1350,
            "y":0,
            "w":75,
            "h":75
         },
         "rotated":false,
         "trimmed":false,
         "spriteSourceSize":{
            "x":0,
            "y":0,
            "w":75,
            "h":75
         },
         "sourceSize":{
            "w":75,
            "h":75
         },
         "duration":100
      },
      "bfPixel UP 0000":{
         "frame":{
            "x":1425,
            "y":0,
            "w":75,
            "h":75
         },
         "rotated":false,
         "trimmed":false,
         "spriteSourceSize":{
            "x":0,
            "y":0,
            "w":75,
            "h":75
         },
         "sourceSize":{
            "w":75,
            "h":75
         },
         "duration":100
      },
      "bfPixel UP 0001":{
         "frame":{
            "x":1500,
            "y":0,
            "w":75,
            "h":75
         },
         "rotated":false,
         "trimmed":false,
         "spriteSourceSize":{
            "x":0,
            "y":0,
            "w":75,
            "h":75
         },
         "sourceSize":{
            "w":75,
            "h":75
         },
         "duration":100
      }
   },
   "meta":{
      "app":"http://www.aseprite.org/",
      "version":"1.2.21-dev",
      "image":"example-bf.png",
      "format":"RGBA8888",
      "size":{
         "w":1575,
         "h":75
      },
      "scale":"1",
      "frameTags":[
         {
            "name":"DOWN MISS",
            "from":0,
            "to":1,
            "direction":"forward"
         },
         {
            "name":"DOWN",
            "from":2,
            "to":3,
            "direction":"forward"
         },
         {
            "name":"IDLE",
            "from":4,
            "to":8,
            "direction":"forward"
         },
         {
            "name":"LEFT MISS",
            "from":9,
            "to":10,
            "direction":"forward"
         },
         {
            "name":"LEFT",
            "from":11,
            "to":12,
            "direction":"forward"
         },
         {
            "name":"RIGHT MISS",
            "from":13,
            "to":14,
            "direction":"forward"
         },
         {
            "name":"RIGHT",
            "from":15,
            "to":16,
            "direction":"forward"
         },
         {
            "name":"UP MISS",
            "from":17,
            "to":18,
            "direction":"forward"
         },
         {
            "name":"UP",
            "from":19,
            "to":20,
            "direction":"forward"
         }
      ],
      "layers":[
         {
            "name":"Sprite Sheet",
            "opacity":255,
            "blendMode":"normal"
         }
      ],
      "slices":[
         
      ]
   }
}
```

**Comment by Geokureli** [comment]
literally seems the same whenever I check out specific parts. I need you to give me an specific example of something this code does that I can't do with `fromTexturePackerJson`

**Comment by MAJigsaw77** [comment]
Aseprite has slices thing that I couldn't done right now, but if I will done them, `texture packers json` will not be able too do it

**Comment by MAJigsaw77** [comment]
Or the duration

**Comment by MAJigsaw77** [comment]
@Geokureli I'll make this I draft until we made the last things for it, ok?

**Comment by Geokureli** [comment]
@MAJigsaw77 just wanna weigh in, the ability to have per-frame durations in atlases would be awesome! I don't see any more features available when I compare the jsons, so please list them out. But durations alone would be worth the effort, if you plan on revisiting this

**Comment by MAJigsaw77** [comment]
k

**Pull request closed without merging**